engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/GatherDiagnostics.scala

```diff
@@ -67,7 +67,7 @@ case object GatherDiagnostics extends IRPass {
       case x                  => x.diagnostics.toList
     }.flatten
     DiagnosticsMeta(
-      diagnostics.distinctBy(d => (d.location, d.message))
+      diagnostics.distinctBy(d => d)
```

Comment:
This code eliminates duplicate diagnostics based on their `location` and `message`. Messages are composed from an array of keys and a string template that concatenates them. To achieve the same without concatenation we can compare on `class`, `location` and `diagnosticKeys`.